contracts/router/quoter/SwapQuoterV2.sol

```diff
@@ -0,0 +1,380 @@
+    /// Will only consider the whitelisted pool for `tokenIn`
+    function _isConnected(
+        uint256 actionMask,
+        address tokenIn,
+        address tokenOut
+    ) internal view returns (bool) {
+        // If token addresses match, no action is required whatsoever.
+        if (tokenIn == tokenOut) {
+            return true;
+        }
+        // Check if ETH <> WETH (Action.HandleEth) could fulfill tokenIn -> tokenOut request.
+        if (Action.HandleEth.isIncluded(actionMask) && _isEthAndWeth(tokenIn, tokenOut)) {
+            return true;
+        }
+        TypedPool memory bridgePool = _bridgePools[tokenIn];
+        // Do nothing, if tokenIn doesn't have a whitelisted pool
+        if (bridgePool.pool == address(0)) return false;
+        if (bridgePool.poolType == PoolType.Default) {
+            // Check if Default Pool could fulfill tokenIn -> tokenOut request.
+            return _isConnectedViaDefaultPool(actionMask, bridgePool.pool, tokenIn, tokenOut);
+        } else {
+            // Check if Linked Pool could fulfill tokenIn -> tokenOut request.
+            return _isConnectedViaLinkedPool(actionMask, bridgePool.pool, tokenIn, tokenOut);
+        }
+    }
+
+    /// @dev Returns the SwapQuery struct that could be used to fulfill `tokenIn -> tokenOut` request.
+    /// - Will check all liquidity pools, if `actionMask` is set to the full set of actions.
+    /// - Will only check the whitelisted pool for `tokenIn` otherwise.
+    /// > Only populates the `minAmountOut` and `rawParams` fields, unless no trade path is found between the tokens.
+    /// > Other fields are supposed to be populated in the caller function.
+    function _getAmountOut(
+        uint256 actionMask,
+        address tokenIn,
+        address tokenOut,
+        uint256 amountIn
+    ) internal view returns (SwapQuery memory query) {
+        // If token addresses match, no action is required whatsoever.
+        if (tokenIn == tokenOut) {
+            query.minAmountOut = amountIn;
+            // query.rawParams is "", indicating that no further action is required
+            return query;
+        }
+        // Check if ETH <> WETH (Action.HandleEth) could fulfill tokenIn -> tokenOut request.
+        _checkHandleETHQuote(actionMask, tokenIn, tokenOut, amountIn, query);
+        // Check if this is a request for an origin swap.
+        // These are given with the tokenIn.actionMask set to the full set of actions.
+        if (actionMask != ActionLib.allActions()) {
+            // This is a request for a destination swap. Only whitelisted pool for `tokenIn` is considered.
+            TypedPool memory bridgePool = _bridgePools[tokenIn];
+            _checkPoolQuote(actionMask, bridgePool, tokenIn, tokenOut, amountIn, query);
+            return query;
+        }
+        unchecked {
+            // If this is a request for an origin swap, check all available pools
```

Comment:
What's the difference between a default pool and a bridge pool?

Reply:
Seems like you're duplicating checks on either default or linked pools as bridge pools are one or another from `_checkPoolQuote()`?

Reply:
> What's the difference between a default pool and a bridge pool?
These are apples and oranges in a way.

SwapQuoterV2 works with two types of pools right now: Default Pools (defined by `IDefaultPool` interface) and Linked Pools (defined by `ILinkedPool` interface).

All these pools are also separated into two categories:
- Pools that could be used only for "origin swap" in a "origin swap + bridge + destination swap" bridging workflow.
  - These are stored as two sets `_defaultPools` and `_linkedPools`
- Pools that could be used both for "origin swap" and "destination swap". Though for destination swap only one pool could be used per bridge token, so these pools are stored in a mapping from a bridge token to a "bridge pool" `_bridgePools`. 
  - The "bridge pools" could be either Default Pools or Linked Pools.

In other words, any of the supported pools could be used to perform a swap on origin chain, but only one pool could be used for swap on destination chain for a given bridge token.

Reply:
> Seems like you're duplicating checks on either default or linked pools as bridge pools are one or another from _checkPoolQuote()?

@mikeyrf I think the culprit here is the not-so-explicit naming on my part. 724128f473f74f4c30d047ac9b84888c6e1bc198 should clarify which pools are "origin-only". In other words, values from `_bridgePools` are not stored in `_defaultPools` and `_linkedPools` (which are now called `_originDefaultPools` and `_originLinkedPools`)